packages/plugin-web-attribution-browser/src/web-attribution.ts

```diff
@@ -27,22 +36,91 @@ export const webAttributionPlugin: CreateWebAttributionPlugin = function (option
         storage.get(storageKey),
       ]);
 
-      const isEventInNewSession = isNewSession(config.sessionTimeout, config.lastEventTime);
+      // Check if the most recent session ID is expired
+      const isMostRecentSessionExpired = isSessionExpired(config.sessionTimeout, config.lastEventTime);
+
+      // Check if attribution event will be tracked
+      if (isNewCampaign(currentCampaign, previousCampaign, pluginConfig, isMostRecentSessionExpired)) {
+        // Set default session ID to be the most recent session ID
+        let sessionId = config.sessionId ?? -1;
 
-      if (isNewCampaign(currentCampaign, previousCampaign, pluginConfig, isEventInNewSession)) {
-        if (pluginConfig.resetSessionOnNewCampaign) {
-          amplitude.setSessionId(Date.now());
-          config.loggerProvider.log('Created a new session for new campaign.');
+        // Check if most recent session ID is expired OR if `resetSessionOnNewCampaign` set to true
+        // If yes, set a new session ID
+        if (isMostRecentSessionExpired || pluginConfig.resetSessionOnNewCampaign) {
+          sessionId = Date.now();
+          amplitude.setSessionId(sessionId);
+
+          if (pluginConfig.resetSessionOnNewCampaign) {
+            config.loggerProvider.log('Created a new session for new campaign.');
+          }
         }
-        config.loggerProvider.log('Tracking attribution.');
+
+        // Create campaign event
         const campaignEvent = createCampaignEvent(currentCampaign, pluginConfig);
+        campaignEvent.session_id = sessionId;
+        // Cache campaign event with its associated session ID as key
+        campaignPerSession[sessionId] = campaignEvent;
+        // Additionally, track the same event
         amplitude.track(campaignEvent);
+
         void storage.set(storageKey, currentCampaign);
       }
     },
 
-    execute: async (event: Event) => event,
+    execute: async (event: Event) => {
+      if (event.session_id) {
+        // Check a campaign event cache exists for session ID
+        const campaignEvent = campaignPerSession[event.session_id];
+        if (campaignEvent) {
+          // If yes, merge first seen event with with campaign event.
+          // It is possible that `event` is equal to `campaignEvent`
+          // when `campaignEvent` is what is first passed to `execute`.
```

Comment:
Not sure I understand. We clear the `campaignPerSession[event.session_id]` after retrieving it. How can event be equal to campaign event?

Reply:
This happens when the first event seen is also the identify event. This may happen because we do 1) caching via `campaignPerSession` and 2) explicitly tracking. The caching strategy ensures that it can be enriched on the first event seen and the explicit tracking strategy ensures that the event is enqueued. To handle potentially duplicate user props, we drop the campaign event if it is no longer in the cache.